src/prted/pmix/pmix_server_dyn.c

```diff
@@ -179,6 +179,7 @@ static void spawn(int sd, short args, void *cbdata)
         PMIX_DATA_BUFFER_RELEASE(buf);
         goto callback;
     }
+    PMIX_DATA_BUFFER_RELEASE(buf);
```

Comment:
Let me look into this a bit - I concur with your analysis, but that means every message sent by PRRTE is currently being leaked, so we need a more global solution. I had intended to do a thorough scrub/review of the messaging system - just not right this moment 😄

Reply:
Pondering this a bit - the "right" solution is to remove that copy_payload and just pass the buffer. This would eliminate unnecessary data duplication and improve the performance of the messaging system.  We would need to ensure that malloc'd buffers were always being passed to `PRTE_RML_SEND` or else we'd crash when the buffer was subsequently released.

Is that too big an effort for you to dig into? If not, it would be appreciated - but I certainly would understand a "pass" on your side 😄

Reply:
Alright. I will try to get this done :)

Reply:
Much appreciated! Please holler if you have any questions or hit any obstacles.

Reply:
My apologies - I can't help but wonder if I have asked you to do something that will descend into a rathole. Perhaps a more reasonable intermediate step would be to do a simple `grep` or `ack` of the code base to find all the places where we send a message, and then just ensure that we release the buffer there as you have done here? This will plug the leaks for the v3.0 release, and that is good enough for now.

Here is what I found:
```shell
src/tools/prted/prted.c
504:        PRTE_RML_SEND(ret, PRTE_PROC_MY_PARENT->rank, &pbuf,
696:    PRTE_RML_SEND(ret, PRTE_PROC_MY_NAME->rank, &buffer, PRTE_RML_TAG_PRTED_CALLBACK);
902:        PRTE_RML_SEND(ret, PRTE_PROC_MY_PARENT->rank, mybucket,

src/runtime/prte_data_server.c
450:                PRTE_RML_SEND(rc, req->proxy.rank, reply, PRTE_RML_TAG_DATA_CLIENT);
869:    PRTE_RML_SEND(rc, sender->rank, answer, PRTE_RML_TAG_DATA_CLIENT);

src/rml/rml.h
78:#define PRTE_RML_SEND(_r, r, b, t)                              \
207:#define PRTE_RML_SEND_COMPLETE(m)                                                             \

src/rml/rml_base_msg_handlers.c
181:            PRTE_RML_SEND(rc, msg->sender.rank, &buffer,

src/prted/prted_comm.c
546:        PRTE_RML_SEND(ret, sender->rank, answer, PRTE_RML_TAG_TOPOLOGY_REPORT);
656:        PRTE_RML_SEND(ret, PRTE_PROC_MY_HNP->rank, answer, PRTE_RML_TAG_STACK_TRACE);

src/prted/pmix/pmix_server_pub.c
198:    PRTE_RML_SEND(rc, target->rank, xfer, PRTE_RML_TAG_DATA_SERVER);

src/prted/pmix/pmix_server_dyn.c
175:    PRTE_RML_SEND(rc, PRTE_PROC_MY_HNP->rank, buf, PRTE_RML_TAG_PLM);

src/prted/pmix/pmix_server_gen.c
604:            PRTE_RML_SEND(rc, PRTE_PROC_MY_HNP->rank,
739:        PRTE_RML_SEND(rc, PRTE_PROC_MY_HNP->rank, buf,

src/prted/pmix/pmix_server.c
977:    PRTE_RML_SEND(prc, remote->rank, reply, PRTE_RML_TAG_DIRECT_MODEX_RESP);
1040:    PRTE_RML_SEND(prc, req->proxy.rank, reply, PRTE_RML_TAG_DIRECT_MODEX_RESP);

src/prted/pmix/pmix_server_fence.c
178:    PRTE_RML_SEND(prc, req->proxy.rank, reply, PRTE_RML_TAG_DIRECT_MODEX_RESP);
359:    PRTE_RML_SEND(rc, dmn->name.rank, buf, PRTE_RML_TAG_DIRECT_MODEX);

src/mca/plm/ssh/plm_ssh_module.c
297:            PRTE_RML_SEND(rc, PRTE_PROC_MY_HNP->rank, buf, PRTE_RML_TAG_REPORT_REMOTE_LAUNCH);
885:        PRTE_RML_SEND(ret, PRTE_PROC_MY_HNP->rank, buf, PRTE_RML_TAG_REPORT_REMOTE_LAUNCH);

src/mca/plm/base/plm_base_launch_support.c
991:    PRTE_RML_SEND(rc, jdata->originator.rank, answer, PRTE_RML_TAG_LAUNCH_RESP);
1693:                    PRTE_RML_SEND(ret, dptr->name.rank, relay, PRTE_RML_TAG_DAEMON);
1770:                PRTE_RML_SEND(ret, dname.rank, relay, PRTE_RML_TAG_DAEMON);

src/mca/plm/base/plm_base_receive.c
189:        PRTE_RML_SEND(ret, sender->rank, answer, PRTE_RML_TAG_JOBID_RESP);
357:        PRTE_RML_SEND(ret, sender->rank, answer, PRTE_RML_TAG_LAUNCH_RESP);

src/mca/errmgr/prted/errmgr_prted.c
213:    PRTE_RML_SEND(rc, PRTE_PROC_MY_HNP->rank, alert, PRTE_RML_TAG_PLM);
306:    PRTE_RML_SEND(rc, PRTE_PROC_MY_HNP->rank, alert, PRTE_RML_TAG_PLM);
514:            PRTE_RML_SEND(rc, PRTE_PROC_MY_HNP->rank, alert, PRTE_RML_TAG_PLM);
627:            PRTE_RML_SEND(rc, PRTE_PROC_MY_HNP->rank, alert, PRTE_RML_TAG_PLM);
686:        PRTE_RML_SEND(rc, PRTE_PROC_MY_HNP->rank, alert, PRTE_RML_TAG_PLM);

src/mca/filem/raw/filem_raw_module.c
856:    PRTE_RML_SEND(rc, PRTE_PROC_MY_HNP->rank, buf, PRTE_RML_TAG_FILEM_BASE_RESP);

src/mca/filem/base/filem_base_receive.c
207:    PRTE_RML_SEND(rc, sender->rank, answer, PRTE_RML_TAG_FILEM_BASE_RESP);
297:    PRTE_RML_SEND(rc, sender->rank, answer, PRTE_RML_TAG_FILEM_BASE_RESP);

src/mca/oob/tcp/oob_tcp_sendrecv.c
251:                    PRTE_RML_SEND_COMPLETE(msg->msg);
266:                PRTE_RML_SEND_COMPLETE(msg->msg);

src/mca/oob/base/oob_base_stubs.c
61:        PRTE_RML_SEND_COMPLETE(msg);
94:                    PRTE_RML_SEND_COMPLETE(msg);
100:                PRTE_RML_SEND_COMPLETE(msg);
143:                PRTE_RML_SEND_COMPLETE(msg);
188:            PRTE_RML_SEND_COMPLETE(msg);
201:        PRTE_RML_SEND_COMPLETE(msg);

src/mca/iof/hnp/iof_hnp_send.c
111:    PRTE_RML_SEND(rc, host->rank, buf, PRTE_RML_TAG_IOF_PROXY);

src/mca/iof/prted/iof_prted_receive.c
69:    PRTE_RML_SEND(rc, PRTE_PROC_MY_HNP->rank, buf, PRTE_RML_TAG_IOF_HNP);

src/mca/iof/prted/iof_prted_read.c
169:    PRTE_RML_SEND(rc, PRTE_PROC_MY_HNP->rank, buf, PRTE_RML_TAG_IOF_HNP);

src/mca/state/dvm/state_dvm.c
704:        PRTE_RML_SEND(rc, PRTE_PROC_MY_NAME->rank, buf, PRTE_RML_TAG_DATA_SERVER);

src/mca/state/prted/state_prted.c
298:        PRTE_RML_SEND(rc, PRTE_PROC_MY_HNP->rank, alert, PRTE_RML_TAG_PLM);
436:            PRTE_RML_SEND(rc, PRTE_PROC_MY_HNP->rank, alert, PRTE_RML_TAG_PLM);
538:            PRTE_RML_SEND(rc, PRTE_PROC_MY_HNP->rank, alert, PRTE_RML_TAG_PLM);

src/mca/state/base/state_base_fns.c
448:    PRTE_RML_SEND(rc, prte_pmix_server_globals.server.rank,

src/mca/ess/base/ess_base_std_prted.c
629:    PRTE_RML_SEND(rc, PRTE_PROC_MY_NAME->rank, cmd, PRTE_RML_TAG_DAEMON);

src/mca/grpcomm/direct/grpcomm_direct.c
102:    PRTE_RML_SEND(rc, PRTE_PROC_MY_HNP->rank, buf, PRTE_RML_TAG_XCAST);
170:    PRTE_RML_SEND(rc, PRTE_PROC_MY_NAME->rank, relay,
355:            PRTE_RML_SEND(rc, PRTE_PROC_MY_PARENT->rank, reply,
579:            PRTE_RML_SEND(ret, nm->rank, rlycopy, PRTE_RML_TAG_XCAST);
```
It's a fairly long list, but I suspect a faster solution. Make sense?

Reply:
Totally up to you as to which way to go. The CI will test them a reasonable amount, and the OMPI folks will bang them pretty well too. Still, it is a higher risk - but definitely better in the long run if you already have it.

Reply:
Yes, I have the first way. I will put it up in some time.